Bug: force_clear_untriaged bypasses FSM, leaves legacy=NonEmpty → worker crash loop

**Symptom**

`FidoCanCode/home` worker crash-loops with the handler-preemption oracle firing every 5–7 minutes:

```
AssertionError: handler_preemption FSM: WorkerTurnStart rejected in state
legacy=LegacyNonEmpty,durable=DurableEmpty,interrupt=InterruptWasRequested
for repo 'FidoCanCode/home'
```

Watchdog auto-restarts after each crash, but the registry-level FSM state survives the worker restart, so the very next `assert_worker_turn_ok` fires the same assertion again. Worker makes no progress.

**Root cause**

`registry.force_clear_untriaged` (`src/fido/registry.py:653-675`) resets the Python-side count and the drained event but **never touches the FSM**:

```python
def force_clear_untriaged(self, repo_name: str) -> int:
    with self._untriaged_lock:
        cleared = self._untriaged.get(repo_name, 0)
        if cleared <= 0:
            return 0
        self._untriaged[repo_name] = 0
        ev = self._untriaged_drained.get(repo_name)
        if ev is not None:
            ev.set()
    log.warning(...)
    return cleared
```

The FSM dict `_preemption_fsm_states` keeps its `legacy=LegacyNonEmpty` because no `HandlerDone` event is fired for the cleared holds. The Python count says "drained, all good"; the FSM says "still N legacy webhooks pending."

This is exactly the failure mode CLAUDE.md "Rocq-modeled coordination boundary" calls out — mutations to authoritative state must route through the modeled transition, not directly. The backstop introduced in #1281 fixed half the state and bypassed the other half.

**Reproduction trace** (from journald):

```
23:27:56 [home] FSM …NotRequested →LegacyNonEmpty,DurableNonEmpty,InterruptNotRequested via WebhookArrives
23:27:58 [home] FSM LegacyNonEmpty,DurableNonEmpty,InterruptNotRequested →LegacyNonEmpty,DurableEmpty,InterruptNotRequested via DurableDemandDrained
23:33:30 AssertionError: WorkerTurnStart rejected in state LegacyNonEmpty,DurableEmpty,InterruptWasRequested
…watchdog restart…
23:40:12 same assertion
…watchdog restart…
23:41:00 same assertion
```

Note the ~5-minute gaps between webhook-arrival and crash — that's the `_INBOX_DRAIN_TIMEOUT_SECONDS = 300.0` cap from `worker.py:101`. The worker is waiting on `wait_for_inbox_drain`, timing out, calling `force_clear_untriaged`, then immediately calling `assert_worker_turn_ok` which the FSM rejects.

The triggering condition appears to be a webhook handler stuck behind a long-running worker turn (currently 31m on #1310's Dispatcher extraction, holding `hold_for_handler` semantics) so `exit_untriaged` doesn't fire within the timeout.

**Fix sketch**

`force_clear_untriaged` must drain the FSM legacy_demand for each cleared hold. Two shapes:

1. Fire `HandlerDone` `cleared` times inside the existing lock — reuses existing transitions, but `HandlerDone` is rejected from some states (e.g. `legacy=LegacyEmpty`) and would itself crash if the FSM disagrees with the Python count.
2. Add a `LegacyForceCleared` event to `models/handler_preemption.v` that unconditionally sets `legacy_demand → LegacyEmpty` regardless of source state. Cleaner — names the operation in the model, makes the proof obligation explicit. Update extraction; call from `force_clear_untriaged`.

Option 2 is the right shape per CLAUDE.md: the operation is *modeled*, not patched around. The model gets a new transition that reflects the real recovery operation (an external timeout-driven reset of legacy demand without paired HandlerDone events), and the lemma set proves the recovery is safe.

The deeper question is whether `force_clear_untriaged` should exist at all once #1310 / Dispatcher extraction lands — the underlying cause (handler stuck behind a long worker turn) might be fixable by yielding the session lock more aggressively rather than papering over with a 5-minute timeout. But the immediate crash-loop fix is the FSM transition.

**Severity:** crash loop in production. Recommend fixing as a standalone bug, independent of the #1272 god-class shattering ordering on #1260, since it's actively blocking work on `FidoCanCode/home`.

**Refs**
- #1280 / #1281 — prior leak-handling fix that introduced the backstop
- CLAUDE.md "Rocq-modeled coordination boundary"

--- a/src/fido/registry.py
+++ b/src/fido/registry.py
@@ -547,7 +547,14 @@ class WorkerRegistry:
         with self._untriaged_lock:
             old = self._untriaged.get(repo_name, 0)
             self._untriaged[repo_name] = old + 1
-            self._preemption_fsm_transition(repo_name, preemption_fsm.WebhookArrives())
+            if old == 0:
+                # Boolean abstraction in handler_preemption.v: legacy demand
+                # is NonEmpty iff *any* handler is in flight.  Fire the FSM
+                # transition only on the 0→1 edge so the FSM agrees with the
+                # Python count by construction.
+                self._preemption_fsm_transition(
+                    repo_name, preemption_fsm.WebhookArrives()
+                )
             ev = self._untriaged_drained.get(repo_name)
             if ev is None:
                 ev = threading.Event()
@@ -575,19 +582,11 @@ class WorkerRegistry:
                 return
             new = old - 1
             self._untriaged[repo_name] = new
-            new_state = self._preemption_fsm_transition(
-                repo_name, preemption_fsm.HandlerDone()
-            )
-            if new > 0:
-                self._preemption_fsm_states[repo_name] = preemption_fsm.with_legacy(
-                    new_state, preemption_fsm.LegacyNonEmpty()
-                )
-                log.debug(
-                    "preemption[%s]: FSM legacy remains non-empty (count is %d)",
-                    repo_name,
-                    new,
-                )
-            else:
+            if new == 0:
+                # Boolean abstraction: fire HandlerDone only on the 1→0 edge.
+                # The FSM agrees with the Python count by construction, so no
+                # post-transition with_legacy override is needed.
+                self._preemption_fsm_transition(repo_name, preemption_fsm.HandlerDone())
                 ev = self._untriaged_drained.get(repo_name)
                 if ev is not None:
                     ev.set()
@@ -663,6 +662,13 @@ class WorkerRegistry:
             if cleared <= 0:
                 return 0
             self._untriaged[repo_name] = 0
+            # The FSM is on the boolean abstraction: legacy is NonEmpty iff
+            # any handler is in flight.  When count > 0, FSM legacy is
+            # NonEmpty by construction (enter_untriaged fired WebhookArrives
+            # on the 0→1 edge), so HandlerDone is the valid drain event.
+            # Without this transition, the FSM stays at LegacyNonEmpty after
+            # force-clear and the next assert_worker_turn_ok crashes (#1330).
+            self._preemption_fsm_transition(repo_name, preemption_fsm.HandlerDone())
             ev = self._untriaged_drained.get(repo_name)
             if ev is not None:
                 ev.set()

--- a/tests/test_registry.py
+++ b/tests/test_registry.py
@@ -1083,6 +1083,21 @@ class TestUntriagedInbox:
         t.join(timeout=2.0)
         assert unblocked.is_set(), "force_clear should unblock waiter"
 
+    def test_force_clear_drains_fsm_legacy_demand(self) -> None:
+        """Regression for #1330: force_clear must take the FSM through the
+        modeled HandlerDone transition, not just zero the Python count.
+
+        Before the fix, force_clear left ``legacy_demand=LegacyNonEmpty`` while
+        the count said 0, so the next ``assert_worker_turn_ok`` fired the FSM
+        oracle and crashed the worker — repeatedly, in a watchdog-restart loop.
+        """
+        reg = self._reg()
+        reg.enter_untriaged("foo/bar")
+        reg.enter_untriaged("foo/bar")  # count=2; FSM is LegacyNonEmpty
+        reg.force_clear_untriaged("foo/bar")
+        # FSM must be back at LegacyEmpty so the worker can take its next turn.
+        reg.assert_worker_turn_ok("foo/bar")  # must not raise
+
 
 class TestPreemptionFsmOracle:
     """Tests for the handler_preemption FSM oracle wired into WorkerRegistry.
@@ -1168,6 +1183,37 @@ class TestPreemptionFsmOracle:
         reg.enter_untriaged("foo/bar")
         reg.exit_untriaged("foo/bar")
         reg.exit_untriaged("foo/bar")
+        reg.assert_worker_turn_ok("foo/bar")  # must not raise
+
+    def test_partial_exit_keeps_fsm_blocked_until_count_drains(self) -> None:
+        """Boolean-abstraction contract: the FSM agrees with the Python count
+        by construction.  Two enters then one exit must leave the FSM at
+        ``LegacyNonEmpty`` (count is still 1) — no spurious HandlerDone fires
+        on the intermediate edge.
+        """
+        reg = self._reg()
+        reg.enter_untriaged("foo/bar")
+        reg.enter_untriaged("foo/bar")
+        reg.exit_untriaged("foo/bar")  # count=1; FSM still LegacyNonEmpty
+
+        with pytest.raises(AssertionError, match="WorkerTurnStart rejected"):
+            reg.assert_worker_turn_ok("foo/bar")
+
+        reg.exit_untriaged("foo/bar")  # count=0; FSM transitions to Empty
+        reg.assert_worker_turn_ok("foo/bar")  # must not raise
+
+    def test_repeated_enter_is_fsm_idempotent(self) -> None:
+        """Multiple enters at the same repo must not crash the FSM oracle.
+
+        Under the boolean abstraction the FSM transition fires only on the
+        ``0 → 1`` count edge; subsequent enters are no-ops at the FSM level.
+        """
+        reg = self._reg()
+        for _ in range(5):
+            reg.enter_untriaged("foo/bar")
+        # Drain to zero — exactly one HandlerDone should fire on the 1→0 edge.
+        for _ in range(5):
+            reg.exit_untriaged("foo/bar")
         reg.assert_worker_turn_ok("foo/bar")  # must not raise
 
     # ── per-repo isolation ───────────────────────────────────────────────

